thrust/thrust/system/hpx/detail/runtime.h

```diff
@@ -0,0 +1,130 @@
+/*
+ *  Copyright 2008-2025 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*! \file runtime.h
+ *  \brief Implementation of the HPX runtime startup/finalization.
+ */
+
+#pragma once
+
+#include <thrust/detail/config.h>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <hpx/include/run_as.hpp>
+#include <hpx/manage_runtime.hpp>
+
+///////////////////////////////////////////////////////////////////////////////
+// Store the command line arguments in global variables to make them available
+// to the startup code.
+
+#if defined(linux) || defined(__linux) || defined(__linux__)
+
+int __argc = 0;
+char** __argv = nullptr;
```

Comment:
this will result in multiple definition errors.. should we inline or move to source file?

Reply:
Inlining might be the easiest way to solve this. However, will the `__attribute__((section(".preinit_array")))` work in this case?

Reply:
Inlining might be the easiest way to solve this.

Reply:
inlining everything works, but now i'm seeing pure virtual function calls during finalization

Reply:
using `thread_local` ensures the desired execution order.. don't know why, taken from HPX's init_globally example

Reply:
I traced this down, turns out the default execution agent is `thread_local`.